pkg/reconciler/pipelinerun/pipelinerun_test.go

```diff
@@ -6607,27 +5918,55 @@ func TestReconcilePipeline_FinalTasks(t *testing.T) {
 					t.Errorf("Expected PipelineRun status to be failed, but was %v for %s",
 						reconciledRun.Status.GetCondition(apis.ConditionSucceeded), tt.name)
 				}
-				if d := cmp.Diff(reconciledRun.Status.TaskRuns, tt.expectedTaskRuns); d != "" {
-					t.Fatalf("Expected PipelineRunTaskRun status to match TaskRun(s) status, but got a mismatch for %s: %s", tt.name, d)
-				}
 			} else if tt.pipelineRunStatusUnknown {
 				// This PipelineRun should still be running and the status should reflect that
 				if !reconciledRun.Status.GetCondition(apis.ConditionSucceeded).IsUnknown() {
 					t.Errorf("Expected PipelineRun status to be unknown (running), but was %v for %s",
 						reconciledRun.Status.GetCondition(apis.ConditionSucceeded), tt.name)
 				}
-				if d := cmp.Diff(reconciledRun.Status.ChildReferences, tt.expectedChildReferences); d != "" {
-					t.Fatalf("Expected PipelineRunTaskRun status to match TaskRun(s) status, but got a mismatch for %s: %s", tt.name, d)
-				}
 			}
 
-			if d := cmp.Diff(reconciledRun.Status.TaskRuns, tt.expectedTaskRuns); d != "" {
-				t.Fatalf("Expected PipelineRunTaskRun status to match TaskRun(s) status, but got a mismatch for %s: %s", tt.name, d)
+			if d := cmp.Diff(reconciledRun.Status.ChildReferences, tt.expectedChildReferences, cmpopts.SortSlices(lessChildReferences)); d != "" {
+				t.Fatalf("Expected PipelineRunTaskRun status to match ChildReferences status, but got a mismatch for %s: %s", tt.name, d)
 			}
+
+			checkTaskRunStatusFromChildRefs(prt.TestAssets.Ctx, t, namespace, clients, reconciledRun.Status.ChildReferences, tt.expectedTaskRuns)
 		})
 	}
 }
 
+// checkTaskRunStatusFromChildRefs checks the status of taskruns from ChildReferences to be expected.
+func checkTaskRunStatusFromChildRefs(ctx context.Context, t *testing.T, namespace string, clients test.Clients,
+	childRefs []v1beta1.ChildStatusReference, expectedTaskRuns map[string]*v1beta1.PipelineRunTaskRunStatus) {
+	t.Helper()
+	if len(expectedTaskRuns) == 0 {
+		return
+	}
+
+	for _, childRef := range childRefs {
+		if childRef.Kind != "TaskRun" {
+			continue
+		}
+		trName := childRef.Name
+
+		trFromChildRef, err := clients.Pipeline.TektonV1beta1().TaskRuns(namespace).Get(ctx, trName, metav1.GetOptions{})
+		if err != nil {
+			t.Fatalf("Failure to get TaskRun from ChildReference %s: %s", childRef.Name, err)
+		}
+
+		if expectedTaskRunStatus, ok := expectedTaskRuns[trName]; !ok {
+			t.Fatalf("Expected not to have TaskRun status from ChildReferences %s", trName)
+		} else if d := cmp.Diff(*expectedTaskRunStatus.Status, trFromChildRef.Status); d != "" {
+			t.Fatalf("Expected TaskRun Status to match ChildReferences TaskRun Status, but got a mismatch %s", diff.PrintWantGot(d))
+		}
+		delete(expectedTaskRuns, trName)
```

Comment:
Is this necessary? I prefer to avoid modifying inputs unless the function is designed to do so

Reply:
Thanks for the suggestion, I delete it so as to makes sure that all expected status are compared.

I changed to using a var for counting if that is better.